caikit/core/data_model/dataobject.py

```diff
@@ -128,13 +129,15 @@ class DataObjectBase(DataBase, metaclass=_DataObjectBaseMetaClass):
     """
 
 
-def dataobject(*args, **kwargs) -> Callable[[Type], Type[DataBase]]:
+_DataObjectBaseT = TypeVar("_DataObjectBaseT", bound=DataObjectBase)
+
+
+def dataobject(*args, **kwargs) -> Callable[[_DataObjectBaseT], Type[_DataObjectBaseT]]:
```

Comment:
Interesting. I think this original type hint was from before I introduced `DataObjectBase` and the decorator would inject `DataBase` as a base class. now that we require the input class to explicitly inherit from `DataObjectBase`, I think the input argument should also be `Type[_DataObjectBaseT]` and in fact it _might_ be just ok with `Type[DataObjectBase]`?

Reply:
I also wonder if the funkiness around needing to add `@dataclass` is because this type hint should technically include a `Union[]` to itself for the special case handling when it's called with no arguments ([line 223](https://github.com/caikit/caikit/pull/271/files#diff-96758051650f0b4a3b80930a8851adc1c588bf8d4d8c22bfe7e4751a9f094e71R238))

Reply:
@gabe-l-hart said:
>  think the input argument should also be Type[_DataObjectBaseT]

Hmm, ok, done.  I'm a little unclear at what `Type[C]` does exactly.   Doesn't the TypeVar already do the trick that  `Type[C]` does, i.e. ["while C (when used as an annotation) refers to instances of class C, Type[C] refers to subclasses of C."](https://peps.python.org/pep-0484/#the-type-of-class-objects)?  If I don't wrap `_DataObjectBaseT` at all with `Type`, it seems to make no difference.  

> and in fact it might be just ok with Type[DataObjectBase]

Seems to be, though I'm a little surprised.  I thought the whole point was the input class defined the output class.  If we define input as `Type[DataObjectBase]` , how does it know that `_DataObjectBaseT` is `TextLine`, for example?  I thought there was some sort of co-constraint thing going on with input and output.

Checking it in for now with

```python
Callable[[_DataObjectBaseT], _DataObjectBaseT]
```

I wish I had a bit better way of playing with type inference.

Reply:
@gabe-l-hart also said:
> I also wonder if the funkiness around needing to add @dataclass is because this type hint should technically include a Union[] to itself for the special case handling when it's called with no arguments ([line 223](https://github.com/caikit/caikit/pull/271/files#diff-96758051650f0b4a3b80930a8851adc1c588bf8d4d8c22bfe7e4751a9f094e71R238))

I've no idea what that would look like, or what you mean by "itself".  It seems to me that doing argument completion for dataclasses without initializers have to be solely a matter of the tooling being able to understand that they are dataclasses?

Reply:
Yeah, this is definitely a game of reverse engineering what the IDE is doing, but it could either be using `@dataclass` as a keyword directly or it could be knowledgable about all the different flavors of decorators (no-call, zero-arg, with-arg) and what their type hint signatures are.